probe-rs/src/debug/function_die.rs

```diff
@@ -26,76 +26,72 @@ pub(crate) struct FunctionDie<'abbrev, 'unit, 'unit_info, 'debug_info> {
 impl<'debugunit, 'abbrev, 'unit: 'debugunit, 'unit_info, 'debug_info>
     FunctionDie<'abbrev, 'unit, 'unit_info, 'debug_info>
 {
+    /// Create a new function DIE reference.
     pub(crate) fn new(
-        die: FunctionDieType<'abbrev, 'unit>,
+        die: Die<'abbrev, 'unit>,
         unit_info: &'unit_info UnitInfo<'debug_info>,
-    ) -> Option<Self> {
+    ) -> Self {
         let tag = die.tag();
 
         match tag {
-            gimli::DW_TAG_subprogram => Some(Self {
+            gimli::DW_TAG_subprogram => Self {
                 unit_info,
                 function_die: die,
                 abstract_die: None,
                 low_pc: 0,
                 high_pc: 0,
                 frame_base: None,
-            }),
+            },
             other_tag => {
-                tracing::error!("FunctionDie has to has to have Tag DW_TAG_subprogram, but tag is {:?}. This is a bug, please report it.", other_tag.static_string());
-                None
+                unreachable!("FunctionDie has to has to have Tag DW_TAG_subprogram, but tag is {:?}. This is a bug, please report it.", other_tag.static_string())
             }
         }
     }
 
+    /// Creates a new inlined function DIE reference.
     pub(crate) fn new_inlined(
-        concrete_die: FunctionDieType<'abbrev, 'unit>,
-        abstract_die: FunctionDieType<'abbrev, 'unit>,
+        concrete_die: Die<'abbrev, 'unit>,
+        abstract_die: Die<'abbrev, 'unit>,
         unit_info: &'unit_info UnitInfo<'debug_info>,
-    ) -> Option<Self> {
+    ) -> Self {
         let tag = concrete_die.tag();
 
-        match tag {
-            gimli::DW_TAG_inlined_subroutine => Some(Self {
-                unit_info,
-                function_die: concrete_die,
-                abstract_die: Some(abstract_die),
-                low_pc: 0,
-                high_pc: 0,
-                frame_base: None,
-            }),
-            other_tag => {
-                tracing::error!("FunctionDie has to has to have Tag DW_TAG_inlined_subroutine, but tag is {:?}. This is a bug, please report it.", other_tag.static_string());
-                None
-            }
+        let gimli::DW_TAG_inlined_subroutine = tag else {
+            unreachable!("FunctionDie has to has to have Tag DW_TAG_inlined_subroutine, but tag is {:?}. This is a bug, please report it.", tag.static_string());
```

Comment:
Why did you change this to `unreachable!`?  The function used to allow for returning a `None`, but now it will panic instead.

Reply:
I thought we catch panics now.

Reply:
We do catch panics in the debugger. If any process panics, it will report the issue to the user, and the debug session ends.

For a case like this, where one of the function identities cannot be resolved, that means the debug experience becomes a complete failure for the user. In most cases, there are valuable debug information in other functions in the stack trace, so even having a partial stack trace can be useful. 

By using `tracing::error`, the error is captured and reported to the user, but they can still continue debugging through other function data. Imagine the top-level function works, but two or three functions deep in the stack trace, there is an error. The error will show, but the user can still evaluate variable values and breakpoints in the top 2 or 3 functions.

If you panic, the session ends.

Does that explain it?

Reply:
Yeah fair, in this case it should not panic :) I'll revert it :)